tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/Tensor.java

```diff
@@ -0,0 +1,256 @@
+    }
+
+    @Override
+    public MemorySegment getSegmentWithHeader() {
+        return segment;
+    }
+
+    @Override
+    public long getNumBytesOfSegmentWithHeader() {
+        return 0;
+    }
+
+    @Override
+    public long getNumBytesOfSegment() {
+        return 0;
+    }
+
+    public long getNumBytesWithoutHeader() {
+        return 0;
+    }
+
+    @Override
+    protected void clear() {
+    }
+
+    public void set(int index, HalfFloat value) {
+        segment.setAtIndex(JAVA_SHORT, baseIndex + index, value.getHalfFloatValue());
+    }
+
+    public void set(int index, float value) {
+        segment.setAtIndex(JAVA_FLOAT, baseIndex + index, value);
+    }
+
+    public HalfFloat get(int index) {
+        short halfFloatValue = segment.getAtIndex(JAVA_SHORT, baseIndex + index);
+        return new HalfFloat(halfFloatValue);
+    }
+
+    public float getFloatValue(int index) {
+        return segment.getAtIndex(JAVA_FLOAT, baseIndex + index);
+    }
+
+    //Example usage: tensorC.set(i, tensorA.get(i, Float.class) + tensorB.get(i, Float.class));
+    public <T> T get(int index, Class<T> type) {
+        if (type.equals(HalfFloat.class)) {
+            short halfFloatValue = segment.getAtIndex(JAVA_SHORT, baseIndex + index);
+            return type.cast(new HalfFloat(halfFloatValue)); // You need to ensure HalfFloat can be cast to T
+        } else if (type.equals(Float.class)) {
+            float floatValue = segment.getAtIndex(JAVA_FLOAT, baseIndex + index);
+            return type.cast(floatValue); // Autoboxing from float to Float
+        }
+        throw new IllegalArgumentException(STR."Unsupported type: \{type}");
+    }
+
+    public void init(HalfFloat value) {
+        assert dType.equals(DType.HALF_FLOAT);
+        for (int i = 0; i < getSize(); i++) {
+            segment.setAtIndex(JAVA_SHORT, baseIndex + i, value.getHalfFloatValue());
+        }
+    }
+
+    public void init(float value) {
+        assert dType.equals(DType.FLOAT);
+        for (int i = 0; i < getSize(); i++) {
+            segment.setAtIndex(JAVA_FLOAT, baseIndex + i, value);
+        }
+    }
+
+    @Override
+    public int getElementSize() {
+        return numberOfElements;
+    }
+
+    public Shape getShape() {
+        return shape;
+    }
+
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+
+        // Basic information
+        sb.append("Tensor(").append(dType.toString()).append(", ").append(shape.toString()).append(")\n");
+
+        int elementsPerLine = 10; // Adjust based on tensor size and preferences
+        for (int i = 0; i < getSize(); i++) {
+            HalfFloat value = get(i);
+
+            sb.append(value.getFloat32());
+            if ((i + 1) % elementsPerLine == 0) {
+                sb.append("\n");
+            } else {
+                sb.append(" ");
+            }
+        }
+        return sb.toString();
+    }
+
+}
```

Comment:
After looking at this class. This only works for FP32 and FP16, but then there are the rest of the classes, such as Int, byte, etc., that are built using the specialized classes.